File: io-hotmoka-network/src/main/java/io/hotmoka/network/models/signatures/MethodSignatureModel.java
package io.hotmoka.network.models.signatures;

import io.hotmoka.beans.annotations.Immutable;
import io.hotmoka.beans.signatures.MethodSignature;
import io.hotmoka.beans.signatures.NonVoidMethodSignature;
import io.hotmoka.beans.signatures.VoidMethodSignature;

/**
 * The model of the signature of a method of a class.
 */
@Immutable
public class MethodSignatureModel extends CodeSignatureModel {

	/**
	 * The name of the method.
	 */
	public final String methodName;

	/**
	 * The return type of the method, if any.
	 */
	public final String returnType;

	/**
	 * Builds the model of the signature of a method.
	 * 
	 * @param parent the original signature to copy
	 */
	public MethodSignatureModel(MethodSignature method) {
		super(method);

		this.methodName = method.methodName;
		if (method instanceof NonVoidMethodSignature)
			returnType = nameOf(((NonVoidMethodSignature) method).returnType);
		else
			returnType = null;
	}

	/**
	 * Yields the method signature corresponding to this model.
	 * 
	 * @return the method signature
	 */
	public MethodSignature toBean() {
		if (returnType == null)
			return new VoidMethodSignature(definingClass, methodName, getFormalsAsTypes());
		else
			return new NonVoidMethodSignature(definingClass, methodName, typeWithName(returnType), getFormalsAsTypes());
	}
}

File: io-hotmoka-network-thin-client/io-hotmoka-network-thin-client-js/src/models/signatures/MethodSignatureModel.ts
import {CodeSignatureModel} from "./CodeSignatureModel";
import {MarshallingContext} from "../../internal/marshalling/MarshallingContext";

/**
 * The model of the signature of a method of a class.
 */
export abstract class MethodSignatureModel extends CodeSignatureModel {
    /**
     * The name of the method.
     */
    methodName: string


    constructor(methodName: string,
                definingClass: string,
                formals: Array<string>) {
        super(definingClass, formals)
        this.methodName = methodName
    }

    public into(context: MarshallingContext): void {
        super.into(context)
        context.writeString(this.methodName)
    }
}